### PR TITLE
Feat: RabbitMQ Operator

### PR DESCRIPTION
**Pull request by ahmedhesham6**

### CODE REVIEW

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+import (
+	"k8s.io/api/core/v1"
+	"guku.io/devx/k8s"
+
+)
+
+#KubeVersion: [=~"^3\\.4\\."]: minor: >=21
+#Values: [=~"^3\\.4\\."]: {
+	global: {
+		imageRegistry: ""
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:12]: this should be a string with a default value

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+		imagePullSecrets: [...v1.#LocalObjectReference]
+		storageClass: ""
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:14]: this should be a string with a default value

Comment (kajogo777) [reply k8s/services/rabbitmq/helm-3.4.cue:14]: `string | *""`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	}
+
+	kubeVersion:       #KubeVersion
+	nameOverride:      string | *""
+	fullnameOverride:  string | *""
+	commonLabels:      k8s.#Labels
+	commonAnnotations: k8s.#Annotations
+	clusterDomain:     "cluster.local"
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:22]: this should be a string with a default value `"cluster.local"`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+#KubeVersion: [=~"^3\\.4\\."]: minor: >=21
+#Values: [=~"^3\\.4\\."]: {
+	global: {
+		imageRegistry: ""
+		imagePullSecrets: [...v1.#LocalObjectReference]
+		storageClass: ""
+	}
+
+	kubeVersion:       #KubeVersion
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:17]: are you sure this is right? we invented the `#KuberVersion` schema for our usage

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	nameOverride:      string | *""
+	fullnameOverride:  string | *""
+	commonLabels:      k8s.#Labels
+	commonAnnotations: k8s.#Annotations
+	clusterDomain:     "cluster.local"
+	extraDeploy: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:23]: this should be a list of something e.g. `[...string]`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+		image: {
+			registry:   string | *"docker.io"
+			repository: string | *"bitnami/rabbitmq-cluster-operator"
+			tag:        string | *"2.2.0-scratch-r7"
+			digest:     string | *""
+			pullPolicy: v1.#enumPullPolicy | *"IfNotPresent"
+			pullSecrets: [...v1.#LocalObjectReference]
+		}
+		replicaCount: uint | *1
+		topologySpreadConstraints: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:54]: what's the schema?

if it's not know we can leave it as anything `[..._]`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+		}
+		containerSecurityContext: v1.#PodSecurityContext | *{
+			enabled:                true
+			runAsUser:              1001
+			runAsNonRoot:           true
+			readOnlyRootFilesystem: true
+		}
+		command: v1.#Command | *[]
+		args:    v1.#Args | *[]
+		hostAliases: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:100]: what's the schema? maybe `[...string]`

if it's not know we can leave it as anything `[..._]`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+		podLabels:             k8s.#Labels
+		podAnnotations:        k8s.#Annotations
+		podAffinityPreset:     string | *""
+		podAntiAffinityPreset: string | *"soft"
+		nodeAffinityPreset:    v1.#Affinity
+		affinity:              v1.#Affinity
+		nodeSelector:          v1.#NodeSelector
+		tolerations: [...v1.#Toleration]
+		updateStrategy: {
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:109]: this should be a schema with a default value, the way it's defined now users will not be able to override this value

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+			type: "RollingUpdate"
+		}
+		priorityClassName: string | *""
+		lifecycleHooks: {}
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:113]: what's the schema? 

the way this is defined now users will not be able to override the value

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+		containerPorts: {
+			metrics: k8s.#Port | *9782
+		}
+		extraEnvVars: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:117]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+		extraEnvVarsCM:     string | *""
+		extraEnvVarsSecret: string | *""
+		extraVolumes: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:120]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+		extraVolumeMounts: []
+		sidecars: []
+		initContainers: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:123]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+		rbac: {
+			create: bool | *true
+		}
+		serviceAccount: #ServiceAccount
+		metrics:        #Metrics
+	}
+
+	msgTopologyOperator: {
+		image: {
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:132]: this cannot be changed by end users, for every string field use `string | *"default value"`

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+			registry:   "docker.io"
+			repository: "bitnami/rmq-messaging-topology-operator"
+			tag:        "1.10.3-scratch-r1"
+			digest:     ""
+			pullPolicy: "IfNotPresent"
+			pullSecrets: [...v1.#LocalObjectReference]
+		}
+		replicaCount: uint | *1
+		topologySpreadConstraints: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:141]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+		containerSecurityContext: v1.#PodSecurityContext | *{
+			enabled:                true
+			runAsUser:              1001
+			runAsNonRoot:           true
+			readOnlyRootFilesystem: true
+		}
+		command:          v1.#Command | *[]
+		args:             v1.#Args | *[]
+		fullnameOverride: string | *""
+		hostAliases: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:193]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+		lifecycleHooks: {}
+		containerPorts: {
+			metrics: k8s.#Port | *8080
+		}
+		extraEnvVars: []
+		extraEnvVarsCM:     string | *""
+		extraEnvVarsSecret: string | *""
+		extraVolumes: []
+		extraVolumeMounts: []
+		sidecars: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:215]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+		initContainers: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:216]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+			ports: k8s.#ServicePort | *{
+				webhook: 443
+			}
+			nodePorts: k8s.#ServicePort | *{
+				http: ""
+			}
+			clusterIP:      string | *""
+			loadBalancerIP: string | *""
+			extraPorts: []
+			loadBalancerSourceRanges: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:234]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+			externalTrafficPolicy: string | *"Cluster"
+			annotations: {}
+			sessionAffinity: string | *"None"
+			sessionAffinityConfig: {}
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:238]: missing schema

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+		}
+	}
+	useCertManager: bool | *false
+}
+
+#ServiceMonitor: {
+	enabled:     bool | *false
+	jobLabel:    k8s.#Labels | *"app.kubernetes.io/name"
+	honorLabels: false
+	selector: {}
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:248]: missing schema

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+			sessionAffinityConfig: {}
+		}
+	}
+	useCertManager: bool | *false
+}
+
+#ServiceMonitor: {
+	enabled:     bool | *false
+	jobLabel:    k8s.#Labels | *"app.kubernetes.io/name"
+	honorLabels: false
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:247]: this value is concrete, use `bool | *false` instead

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+	selector: {}
+	scrapeTimeout: string | *""
+	interval:      string | *""
+	metricRelabelings: []
+	relabelings: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:252]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+	labels: {}
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:253]: missing schema

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+
+	type: string | *"ClusterIP"
+
+	// -- Metrics service port to scrape
+	port:      k8s.#Port | *80
+	nodePorts: k8s.#Port | *{
+		http: ""
+	}
+	clusterIP: string | *""
+	extraPorts: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:281]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+	loadBalancerIP: string | *""
+	loadBalancerSourceRanges: []
```
Comment (kajogo777) [inline k8s/services/rabbitmq/helm-3.4.cue:283]: what's the schema? maybe [...string]

if it's not know we can leave it as anything [..._]

**Review by kajogo777 — CHANGES_REQUESTED** [review]
There are some concrete fields that cannot be changed by end-users, these should be turned into a schema/constraint with a default value.

I suspect there's  a field with the wrong schema that should be updated "kubeVersion"

**Review by kajogo777 — CHANGES_REQUESTED** [review]

File: k8s/services/rabbitmq/helm-3.4.cue [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+	clusterIP: string | *""
+	extraPorts: [...]
+	loadBalancerIP: string | *""
+	loadBalancerSourceRanges: [...]
+	externalTrafficPolicy: string | *"Cluster"
+
+	// -- Additional service annotations
+	annotations: k8s.#Annotations | *{
+		"prometheus.io/scrape": "true"
+		"prometheus.io/port":   #Metrics.service.port
```
[the remark this answers is not in the document]
Comment (ahmedhesham6) [reply k8s/services/rabbitmq/helm-3.4.cue:288]: We want it referenced to a value

[the remark this answers is not in the document]
Comment (kajogo777) [reply k8s/services/rabbitmq/helm-3.4.cue:288]: then just use `port`

**Review by kajogo777 — APPROVED** [review]

**Pull request merged**